### PR TITLE
Xin/support deep levels reserved json name

### PR DESCRIPTION
**Pull request by xin-au**
Supporting deep levels reserved json name. Regarding the issue https://github.com/grpc-ecosystem/grpc-gateway/issues/1187

### CODE REVIEW

**Comment by xin-au** [comment]
@johanbrandhorst , I am working on a test, will push another commit.

**Comment by xin-au** [comment]
@johanbrandhorst , I finalized the PR. It is ready for review.

File: protoc-gen-swagger/genswagger/template_test.go [hunk 1]
```diff
@@ -2369,3 +2402,93 @@ func TestTemplateWithoutErrorDefinition(t *testing.T) {
...
+		fieldName                     string
+		messageNameToFieldsToJSONName map[string]map[string]string
+		fieldNameToType               map[string]string
+	}
+	tests := []struct {
+		name string
+		args args
+		want string
+	}{
+		// TODO: Add test cases.
```
Comment (johanbrandhorst) [inline protoc-gen-swagger/genswagger/template_test.go]: I think you can remove this 😁

File: protoc-gen-swagger/genswagger/template.go [hunk 2]
```diff
@@ -1935,15 +1935,55 @@ func addCustomRefs(d swaggerDefinitionsObject, reg *descriptor.Registry, refs re
...
+		for _, oneField := range msg.GetField()  {
+			fieldNameToJSONName[oneField.GetName()] = oneField.GetJsonName()
+			fieldNameToType[oneField.GetName()] = oneField.GetTypeName()
+		}
+		messageNameToFieldsToJSONName[msg.GetName()] = fieldNameToJSONName
+	}
+	if strings.Contains(fieldName, ".") {
+		fieldNames := strings.Split(fieldName, ".")
+		fieldNamesWithCamelCase := make([]string, 0)
+		for _, oneField := range fieldNames[:len(fieldNames) - 1] {
```
[the remark this answers is not in the document]
Comment (xin-au) [reply protoc-gen-swagger/genswagger/template.go]: Good point! Fixing it now.

File: protoc-gen-swagger/genswagger/template_test.go [hunk 3]
```diff
@@ -1314,6 +1314,37 @@ func generateFieldsForJSONReservedName() []*descriptor.Field {
 	return append(fields, field)
 }
 
+func generateMsgsForJSONReservedName() []*descriptor.Message {
+	result := make([]*descriptor.Message, 0)
+	//The first message, its field is field_abc and its type is NewType
```
[the remark this answers is not in the document]
Comment (xin-au) [reply protoc-gen-swagger/genswagger/template_test.go]: Good catch! Thanks for pointing it out!

File: protoc-gen-swagger/genswagger/template_test.go (hunk 1 above)
Comment (xin-au) [reply protoc-gen-swagger/genswagger/template_test.go]: Good catch! :grin: Fixing it.

**Comment by xin-au** [comment]
@johanbrandhorst , good call! I tested on my side using the @brocaar 's example, it works. 
@johanbrandhorst @brocaar , sorry for any inconvenience it caused. Welcome any feedback.

**Comment by johanbrandhorst** [comment]
Thanks for your contribution!

**Comment by xin-au** [comment]
@johanbrandhorst , anytime!

**Comment by brocaar** [comment]
Thanks @xin-au for the quick response and fix! :+1:

**Comment by xin-au** [comment]
@brocaar , anytime!

**Pull request merged**